### PR TITLE
retry preparing statements when there is a cassandra error

### PR DESCRIPTION
**Pull request by perobertson**
I started discovering new Cassandra Errors being propagated up when updating to v2 of this gem. I traced it down to the fact that preparing the statement can raise most of the same errors as the execution. I also thought it was appropriate to switch to catching the errors specified in the docs. 

[prepare docs](http://docs.datastax.com/en/developer/ruby-driver/3.0/api/cassandra/session/#prepare-instance_method)
[execute docs](http://docs.datastax.com/en/developer/ruby-driver/3.0/api/cassandra/session/#execute-instance_method)

### CODE REVIEW

File: spec/examples/metal/keyspace_spec.rb [hunk 1]
```diff
@@ -126,16 +126,70 @@
...
 
         expect { cequel.execute(statement) }.not_to raise_error
       end
+
+      it "reconnects to cassandra with a new client after execution failed" do
+        allow(cequel.client)
+          .to receive(:execute)
+               .with(->(s){ s.cql == statement},
+                     hash_including(:consistency => cequel.default_consistency))
+          .and_raise(Cassandra::Errors::NoHostsAvailable)
```
Comment (perobertson) [inline spec/examples/metal/keyspace_spec.rb]: This was actually supposed to raise `Cassandra::Errors::ExecutionError`

File: spec/examples/metal/keyspace_spec.rb [hunk 2]
```diff
@@ -126,16 +126,70 @@
...
+    context "without a connection error" do
+      it "executes a CQL query" do
+        expect { cequel.prepare_statement(statement) }.not_to raise_error
+      end
+    end
+
+    context "with a connection error" do
+      it "reconnects to cassandra with a new client after no hosts could be reached" do
+        allow(cequel.client)
+          .to receive(:prepare_statement)
```
Comment (perobertson) [inline spec/examples/metal/keyspace_spec.rb]: the expected receives were supposed to be just `prepare`

**Comment by pezra** [comment]
It makes completel sense that prepare would fail that way. Thanks for the improvement.

**Comment by perobertson** [comment]
looks like a couple fixes didnt get made before my battery died. I can make those changes later tonight when im back online.

File: lib/cequel/metal/keyspace.rb [hunk 3]
```diff
@@ -226,6 +227,32 @@ def execute_with_options(statement, options={})
...
+          client.prepare(cql)
+        rescue Cassandra::Errors::NoHostsAvailable,
+               Cassandra::Errors::ExecutionError
+          clear_active_connections!
+          raise if retries == 0
+          retries -= 1
+          sleep(retry_delay)
+          retry
+        end
+      end
```
Comment (pezra) [inline lib/cequel/metal/keyspace.rb]: Could we extract this and the execute retry logic into a retry method that takes a block?

**Comment by perobertson** [comment]
dried up the retry code and fixed up the tests with the correct exceptions

**Comment by pezra** [comment]
Thanks for the improvement!

**Comment by perobertson** [comment]
Anytime! The next thing Ill probably look into is making this gem thread safe as it would be nice to use sidekiq.

**Comment by pezra** [comment]
Thread safety would be awesome.

**Comment by pezra** [comment]
Released in 2.0.2

**Pull request merged**